File: lib-net/src/main/java/com/llj/lib/net/observer/BaseApiObserver.java
package com.llj.lib.net.observer;

import android.support.annotation.NonNull;

import com.llj.lib.net.response.BaseResponse;

import io.reactivex.SingleObserver;
import io.reactivex.disposables.Disposable;

/**
 * ArchitectureDemo
 * describe:
 * author liulj
 * date 2018/5/7
 */
public abstract class BaseApiObserver<Data> implements SingleObserver<BaseResponse<Data>>, IObserverTag {
    private Disposable mDisposable;
    private int        mTag;

    public BaseApiObserver(int tag) {
        mTag = tag;
    }

    public BaseApiObserver(ITag iTag) {
        if (iTag != null) {
            mTag = iTag.getRequestTag();
        }
    }

    @Override
    public void onSubscribe(@NonNull Disposable d) {
        mDisposable = d;
    }

    @Override
    public void onSuccess(@NonNull BaseResponse<Data> response) {

    }

    @Override
    public void onError(@NonNull Throwable t) {
    }


    @Override
    public void setRequestTag(int tag) {
        this.mTag = tag;
    }

    @Override
    public int getRequestTag() {
        return mTag;
    }

    @Override
    public Disposable getDisposable() {
        return mDisposable;
    }
}

File: lib-net/src/main/java/com/llj/lib/net/observer/BaseApiObserver.kt
package com.llj.lib.net.observer

import com.llj.lib.net.response.BaseResponse

import io.reactivex.SingleObserver
import io.reactivex.disposables.Disposable

/**
 * ArchitectureDemo
 * describe:
 * author liulj
 * date 2018/5/7
 */
abstract class BaseApiObserver<Data> : SingleObserver<BaseResponse<Data>>, IObserverTask {
    private lateinit var mDisposable: Disposable
    private var mTaskId: Int = 0


    constructor(taskId: Int) {
        mTaskId = taskId
    }

    constructor(iTaskId: ITaskId?) {
        mTaskId = iTaskId?.getRequestId() ?: 0
    }

    constructor()

    override fun onSubscribe(d: Disposable) {
        mDisposable = d
    }

    override fun onSuccess(response: BaseResponse<Data>) {

    }

    override fun onError(t: Throwable) {}


    override fun setRequestId(taskId: Int) {
        this.mTaskId = taskId
    }

    override fun getRequestId(): Int {
        return mTaskId
    }

    override fun getDisposable(): Disposable {
        return mDisposable
    }
}
